QUARKUS-2284.md

```diff
@@ -0,0 +1,79 @@
+# QUARKUS-2284 OpenShift Serverless is requesting that we support it when using OpenShift Serverless Functions
+
+Jira: https://issues.redhat.com/browse/QUARKUS-2284
+
+This feature is about productization of OpenShift Serverless Functions and Quarkus Funqy Knative Events Binding extension.
+We should use OpenShift Serverless tooling to generate Quarkus function, build it and deploy it and make sure that Funqy features works as expected.
+Before we can do that, we need to install Knative Eventing operator with our Quarkus QE Framework as we want the Test Suite to work without any prerequisites and the operator starts several pods that would keep running.
+
+## Scope of the testing
+
+Tests should verify:
+- Quarkus function can be created and deployed using the Knative (`kn`) CLI
+- Creating more complex Quarkus function using YAML file and OpenShift CLI (`oc`)
+- Invoking Quarkus functions:
+  - JSON object in the body of the request (HTTP POST request)
+  - Data in the query string (HTTP GET request)
+  - JSON object in the data property (CloudEvent)
+- Quarkus function receives and processes:
+  - bean (f.e. POJO)
+  - `CloudEvent` using the Funqy `@Context` annotation
+  - binary data
+  - primitive type or its wrapper
+- Reading and writing of the attributes of a `CloudEvent`
+- Permitted return types:
+  - `void`
+  - `String`
+  - `byte[]`
+  - primitive type and its wrapper
+  - Javabeans, maps, lists, arrays
+  - `CloudEvents<E>`
+  - `Uni<T>`, where `T` is any of types listed above
+- Quarkus function accepts events from custom resource created by us as well as from Event Sources provided out of the box
+- Cloud Event type and source matching:
+  - default mapping when the type match the function name for the function to trigger
+  - mapping changed by configuration within application.properties
+  - mapping with `@CloudEventMapping` annotation
+- Complex function chain, when one function triggers the next and so on
+- DEV mode and unit testing using RestAssured using:
+  - normal HTTP requests
+  - Cloud Event Binary mode
+  - Structured Mode
+- Multiple functions declared in one application
+- Accessing secrets and config maps from functions
+
+### Impact on test suites and testing automation
+
+* One new module will be added to our Test Suite:
+  * Bare metal tests that cover DEV mode (JVM mode only)
+  * OpenShift tests for Funqy applications created, build and deployed as OpenShift Serverless Functions
+
+### Impact on resources:
+
+- Impact on lab capacity is one medium size executor and up to 40 minutes of execution time
+- Majority of the time will be needed to install the operator, build and deploy the function, while tests themselves should only run for tens of seconds
```

Comment:
Nope, when I build Quarkus function using `kn func`, it takes about 15 minutes (I did it countless times), and deploying using same tool takes about 10 minutes. I acknowledge it could go down to my miserable internet connection, but I don't have such significant issues with uploads when running our TS `-Dopenshift`.

You can try it yoursef:
```
1. oc new-project funqy-mjurc
2. kn func create -l quarkus -t cloudevents knative-events # this depends on your func version, could be `events`
3. kn func build
4. kn func deploy
```

That time even ignores a time required to downloading ` CNCF Cloud Native Buildpacks` image. ATM using `Red Hat Source-to-Image (S2I)` doesn't work for me and I've seen similar issues described in GChat serverless (knative) forum.

Reply:
I've tried it a minutes ago on fresh installation of OCP 4.11 and build took ~ 5 min, deploy ~ 20 min.

Reply:
I changed execution time accordingly.

Reply:
@mjurc re-tested and with YAML/oc way we go now, times are heavily different.